## 창고 정리 
- 창고에 상자가 가로방향으로 일렬로 쌓여 있습니다. 만약 가로의 길이가 7이라면 1열은 높이가 6으로 6개의 상자가 쌓여 있고, 2열은 3개의 상자, 3열은 9개의 상자가 쌓여 있 으며 높이는 9라고 읽는다.
- 창고 높이 조정은 가장 높은 곳에 상자를 가장 낮은 곳으로 이동하는 것을 말한다.
- 가장 높은 곳이나 가장 낮은 곳이 여러곳이면 그 중 아무거나 선택하면 된다.
- 위에 그림을 1회 높이 조정을 하면 다음과 같아진다.
- 창고의 가로 길이와 각 열의 상자 높이가 주어집니다. m회의 높이 조정을 한 후 가장 높은 곳 과 가장 낮은 곳의 차이를 출력하는 프로그램을 작성하세요.

- ▣ 입력설명
    - 첫 번째 줄에 창고 가로의 길이인 자연수 L(1<=L<=100)이 주어집니다.
    - 두 번째 줄에 L개의 자연수가 공백을 사이에 두고 입력됩니다. 각 자연수는 100을 넘지 않습니다
    - 세 번째 줄에 높이 조정 횟수인 M(1<=M<=1,000)이 주어집니다.
- ▣ 출력설명
    - M회의 높이 조정을 마친 후 가장 높은곳과 가장 낮은 곳의 차이를 출력하세요.

- ▣ 입력예제 1
    - 10 
    - 69 50 42 68 76 40 87 14 65 76 81

- ▣ 출력예제 1
    - 20


In [1]:
'''
1. 입력받은 M회 반복
2. 가장 큰 높이의 값과 가장 작은 높이의 값을 찾는다 (인덱스를 바탕으로)
3. 큰 높이의 값에서 -1을 빼고 작은 높이의 값에서 +1 한다 
4. 반복이 끝난다면 다시 가장 큰 높이의 값과 가장 작은 높이의 차를 출력

위의 방법은 뭔가 비효율적인 느낌이 남 
'''

'\n1. 입력받은 M회 반복\n2. 가장 큰 높이의 값과 가장 작은 높이의 값을 찾는다 \n3. 큰 높이의 값에서 -1을 빼고 작은 높이의 값에서 +1 한다 \n4. 반복이 끝난다면 다시 가장 큰 높이의 값과 가장 작은 높이의 차를 출력\n\n'

In [29]:
L = int(input())
# li = [] 

    # li.append(int(input()))
li = list(map(int, input().split()))
M = int(input())

print(L, li, M)

20 [56, 80, 91, 56, 36, 97, 47, 15, 12, 70, 40, 71, 66, 10, 12, 68, 72, 38, 31, 55] 150


In [30]:
li2 = li.copy()
li2

[56,
 80,
 91,
 56,
 36,
 97,
 47,
 15,
 12,
 70,
 40,
 71,
 66,
 10,
 12,
 68,
 72,
 38,
 31,
 55]

In [31]:
for j in range(M):
    # 임시 max, min 값 
    temp_max = max(li2)
    temp_min = min(li2)

    # 반복문을 돌면서 해당 값을 가진 index만 추출 
    for i in range(len(li2)):
        if li2[i] == temp_max:
            temp_max_idx = i
        if li2[i] == temp_min:
            temp_min_idx = i 
    
    li2[temp_max_idx] = li2[temp_max_idx] - 1 
    li2[temp_min_idx] = li2[temp_min_idx] + 1

print(max(li2) - min(li2))

16


### 추천 풀이

In [ ]:
L = int(input())

a = list(map(int, input().split()))

m = int(input())

# 아... 이렇게 하면 더 효율적이네 
a.sort()

for _ in range(m):
    a[0] = a[0] + 1
    a[L-1] = a[L-1] - 1
    a.sort()

print(a[L-1] - a[0])

# sort를 사용하면 내가 적은 코드처럼 값과 index를 선언할 필요가 없음

## 침몰하는 타이타닉(그리디)
- 유럽에서 가장 유명했던 유람선 타이타닉이 침몰하고 있습니다. 유람선에는 N명의 승객이 타고 있습니다. 
- 구명보트를 타고 탈출해야 하는데 타이타닉에 있는 구명보트는 2명 이하로만 탈 수 있 으며, 보트 한 개에 탈 수 있는 총 무게도 M kg 이하로 제한되어 있습니다.
- N명의 승객 몸무게가 주어졌을 때 승객 모두가 탈출하기 위한 구명보트의 최소개수를 출력하는 프로그램을 작성하세요.

- ▣ 입력설명
    - 첫째 줄에 자연수 N(5<=N<=1000)과 M(70<=M<=250)이 주어집니다.
    - 두 번째 줄에 N개로 구성된 몸무게 수열이 주어집니다. 몸무게는 50이상 150이하입니다. 각 승객의 몸무게는 M을 넘지는 않습니다. 즉 탈출을 못하는 경우는 없습니다.
- ▣ 출력설명
    - 첫째 줄에 구명보트의 최소 개수를 출력합니다.

- ▣ 입력예제 1
    - 5 140
    - 90 50 70 100 60
- ▣ 출력예제 1 
    - 3

In [32]:
'''
1. 정렬하여 최소 최대가 한개의 보트에 탈 수 있는지 확인
    1.1 탈 수 있다면 그 값들 없애기 
    1.2 탈 수 없다면 최대만 없애기 
2. 위의 과정을 반복했을 때 남은 사람이 1명이면 +1하고 0명이면 그대로 출력 
'''

'\n1. 정렬하여 최소 최대가 한개의 보트에 탈 수 있는지 확인\n    1.1 탈 수 있다면 그 값들 없애기 \n    1.2 탈 수 없다면 최대만 없애기 \n2. 위의 과정을 반복했을 때 남은 사람이 1명이면 +1하고 0명이면 그대로 출력 \n'

In [39]:
N, M = map(int, input().split())
li = list(map(int, input().split()))
print(N, M)
print(li)

20 120
[68, 72, 30, 105, 55, 115, 36, 67, 119, 111, 95, 24, 25, 80, 55, 85, 75, 83, 21, 81]


In [40]:
cnt = 0 
while True:
    if len(li) == 0:
        break 

    li.sort()

    # 최대, 최소 사람이 함께 탈 수 있다면 count 진행하고 빼기 
    if max(li) + min(li) <= M:
        cnt += 1 
        li.pop()
        li.pop(0)
    # 할 수 없다면 가장 큰 사람만 빼기
    else:
        cnt += 1 
        li.pop()
    
print(cnt)

14


### 추천 풀이

In [ ]:
# 내가 생각한 아이디어랑 비슷하네 (정렬 후 비교 진행하면서 보트 count)

n, limit = map(int, input().split())
p = list(map(int, input().split()))

p.sort()

cnt = 0 

# p가 비어있다면 거짓으로 빠져나오고 비어있지 않다면 빠져나오지 않음 
while p:

    # else에서 만약 1명만 남았는데 pop(0)를 한 후 밑에 또 pop()이 있으면 에러가 나기 때문에 조건절 추가 
    if len(p) == 1:
        cnt += 1
        break 

    # 아 위에서도 sort를 했는데 굳이 max,min 함수 사용했네...
    # 가장 무거운 사람만 보트타고 나갔을 떄 
    if p[0] + p[-1] > limit:
        p.pop()
        cnt += 1 
    
    else:
        p.pop(0)
        p.pop()
        cnt += 1 

print(cnt)


In [ ]:
# list를 pop하게 된다면 앞으로 땡겨지는 연산을 한번 진행하기 때문에 deque 사용하는게 더 효율적임 

from collections import deque

n, limit = map(int, input().split())
p = list(map(int, input().split()))

p.sort()

# deque로 변경 
p = deque(p)

cnt = 0 

while p:

    if len(p) == 1:
        cnt += 1
        break 

    if p[0] + p[-1] > limit:
        p.pop()
        cnt += 1 
    
    else:
        # 왼쪽 추출 
        p.popleft()
        p.pop()
        cnt += 1 

print(cnt)
